scipy/sparse/linalg/dsolve/linsolve.py

```diff
@@ -479,51 +481,63 @@ def spsolve_triangular(A, b, lower=True, overwrite_A=False, overwrite_b=False):
         A = A.copy()
 
     if A.shape[0] != A.shape[1]:
-        raise ValueError('A must be a square matrix but its shape is {}.'.format(A.shape))
+        raise ValueError(
+            'A must be a square matrix but its shape is {}.'.format(A.shape))
 
     A.eliminate_zeros()
     A.sort_indices()
 
     b = np.asanyarray(b)
 
     if b.ndim not in [1, 2]:
-        raise ValueError('b must have 1 or 2 dims but its shape is {}.'.format(b.shape))
+        raise ValueError(
+            'b must have 1 or 2 dims but its shape is {}.'.format(b.shape))
     if A.shape[0] != b.shape[0]:
-        raise ValueError('The size of the dimensions of A must be equal to '
+        raise ValueError(
+            'The size of the dimensions of A must be equal to '
             'the size of the first dimension of b but the shape of A is '
             '{} and the shape of b is {}.'.format(A.shape, b.shape))
 
-    # Init x as copy of b.
+    # Init x as (a copy of) b.
+    x_dtype = np.result_type(A.data, b, np.float)
     if overwrite_b:
-        x = b
+        if np.can_cast(b.dtype, x_dtype, casting='same_kind'):
+            x = b
+        else:
+            raise ValueError(
+                'b has dtype {} and A has dtype {}. Hence b can not store the '
+                'result. But overwrite_b is True. Please set overwrite_b to '
+                'False.'.format(b.dtype, A.data.dtype))
     else:
-        x = b.copy()
+        x = b.astype(x_dtype, copy=True)
 
     # Choose forward or backward order.
     if lower:
         row_indices = range(len(b))
     else:
-        row_indices = range(len(b)-1, -1, -1)
+        row_indices = range(len(b) - 1, -1, -1)
 
     # Fill x iteratively.
     for i in row_indices:
 
         # Get indices for i-th row.
         indptr_start = A.indptr[i]
-        indptr_stop = A.indptr[i+1]
+        indptr_stop = A.indptr[i + 1]
         if lower:
-            A_diagonal_index_row_i = indptr_stop-1
-            A_off_diagonal_indices_row_i = slice(indptr_start,indptr_stop-1)
+            A_diagonal_index_row_i = indptr_stop - 1
+            A_off_diagonal_indices_row_i = slice(indptr_start, indptr_stop - 1)
         else:
             A_diagonal_index_row_i = indptr_start
-            A_off_diagonal_indices_row_i = slice(indptr_start+1,indptr_stop)
+            A_off_diagonal_indices_row_i = slice(indptr_start + 1, indptr_stop)
 
         # Check regularity and triangularity of A.
         if indptr_stop <= indptr_start or A.indices[A_diagonal_index_row_i] < i:
-            raise LinAlgError('A is singular: '
+            raise LinAlgError(
+                'A is singular: '
                 '{}th diagonal is zero!'.format(i))
         if A.indices[A_diagonal_index_row_i] > i:
-            raise LinAlgError('A is no triangular matrix: entry '
+            raise LinAlgError(
+                'A is no triangular matrix: entry '
                 '[{},{}] is not zero!'.format(i, A.indices[A_diagonal_index_row_i]))
```

Comment:
While we're updating these lines, I find error messages with exclamation points to be a bit excessive:

 * "A is singular: diagonal {} is zero."
 * "A is not triangular: A[{}, {}] is nonzero."